Issue 13: QR Codes

Nabushika opened the pull request:
Fixes #13 
## Changes
* Uses `qrcode-generator` crate.
* Cross-service (local/shuttle) router creation, to avoid updates in two places.
* Moved calendar generation to new file `cal.rs` to allow reuse between qr generation and normal function.
* Fixed clippy complaining about a `match` which could have been an `if let`.


## Design
### Decisions
* Exports images as PNG for better compatibility (HTML image embedding didn't work with SVG), plus smaller file sizes (<2KB, compared to ~200KB for SVG).
* QR generation is handled by a new endpoint `/qr`, for continued ease of use with `curl` etc., plus this allows people to share links to images rather than the image itself if desired.
* QR contains embedded calendar event, rather than a link back to the app to generate the event.
* QR code has medium ECC - not personally sure if this is correct, we could go lower for smaller codes.
### UI/UX
QR code is tucked beneath the link to the calendar invite:
![image](https://user-images.githubusercontent.com/65562885/201472728-b71611be-c3da-4723-abb6-cfdd73fb5a75.png)

QR code directly produces a calendar event on both Android and iOS:
Android | iOS
-----------|-------
![image](https://user-images.githubusercontent.com/65562885/201472784-b5e0a968-2037-4454-b6a4-3927dc5e5ec9.png)|![image](https://user-images.githubusercontent.com/65562885/201478028-7213d0ae-021c-4a76-ac39-26b9c922382e.png)

mre commented:
Wow, this is a nice contribution. Thanks a lot @Nabushika.
I tested it locally and it works like a charm. As expected it did create a calendar event on Android.
It's also great that you refactored the router code. Something that I wanted to do as well. 👍 

The QR code has quite a high resolution. I was wondering if e.g. a version 1 or version 2 code would suffice as well.
In any case I think we can tweak that later so it's not a blocker. Just wanted to pick your brain on this before merging it in.

Apart from that the code looks great and the QR code crate is well-maintained, so this is ready to be deployed from my side.

Nabushika commented:
Having a look see if I can clean up the error handling

mre commented:
Good job! Shall I merge and deploy it or would you like to make any other changes before?

Nabushika commented:
As I said, I think I can clean up the error handling a little first, I shouldn't have introduced a new error type since you're using anyhow anyway. If you give me another day that should be enough to clean it up a little?

mre commented:
Oh, sorry. I misread your message. That's totally fine. 👍

Nabushika commented:
I cleaned up the start/end code in `create_calendar` and added some tests to make sure it works correctly - I haven't tested the "now" functionality, was a little unsure how to go about that without creating flaky tests, maybe we'd want to split the calendar function up into two functions, one that takes params plus a "now" parameter, and one that just passes in `Utc::now` as the now?
I also got rid of my custom error type, now preferring to use `anyhow::error` like the rest of your code.

Anyway, I think I'm happy with this now. Let me know if you'd like any other changes, otherwise feel free to merge :)

Nabushika commented:
As for the "high resolution" comment, I believe that's because the whole calendar file is encoded into the QR code, if we made it link to a URL it'd probably be smaller (it grows to fit as you type in more data currently) - but then the user would have to head to the website and maybe download the calendar file before being able to add to their calendar. I think this way has a nicer experience as it comes up directly on the user's phone.

Nabushika commented:
Okay, actually done now, I promise 😂

mre commented on src/cal.rs:
```diff
@@ -43,19 +43,16 @@ pub fn create_calendar(params: HashMap<String, String>) -> Result<Calendar> {
         .map_err(|e| e.context("Invalid duration"))?;
 
-    let (start, end) = match (start, end, duration) {
+    let (start, end) = match (start, end, duration.unwrap_or(DEFAULT_EVENT_DURATION)) {
```
Pretty smart to combine all cases into a `match` statement by the way. 👍

mre commented:
> Having said that, currently ECC is set to medium, meaning 15% of the qr code can be misread. We could change this to low, which drops it to 7% but would make it smaller.

We can always change it later. I think it's fine actually.
I have to say it's kinda fun to change the data and watch the QR code change. 😆 

> I haven't tested the "now" functionality, was a little unsure how to go about that without creating flaky tests, maybe we'd want to split the calendar function up into two functions, one that takes params plus a "now" parameter, and one that just passes in Utc::now as the now?

Good idea. Maybe that's something we could tackle in the future.

> Okay, actually done now, I promise 😂

Haha, no worries. I'll take your word for it. 😉 Merging and deploying.

mre commented:
Oookay the deployment failed with some weird caching issue.
Deleting and recreating the project didn't help.
I asked for help in the Shuttle Discord. Please hold on. ^^

Nabushika commented:
Oh dear
